contentcuration/contentcuration/frontend/shared/views/channel/mixins.js

```diff
@@ -0,0 +1,122 @@
+  authors: 'Authors',
+  providers: 'Providers',
+  aggregators: 'Aggregators',
+  licenses: 'Licenses',
+  copyrightHolders: 'Copyright holders',
+  yes: 'Yes',
+  no: 'No',
+  kindCount: '{kind} ({count})',
+  downloadFilename: '{year}_{month}_Kolibri_Content_Library',
+});
+
+const exportExtensionMap = {
+  csv: 'text/csv',
+  pdf: 'application/pdf',
+};
+
+export const channelExportMixin = {
+  computed: {
+    exportStrings() {
+      return exportStrings;
+    },
+  },
+  mixins: [fileSizeMixin, constantsTranslationMixin],
+  methods: {
+    ...mapActions('channel', ['getChannelListDetails']),
+    _generateDownload(content, extension) {
+      let blob = new Blob([content]);
+      let downloadButton = document.createElement('a');
+      downloadButton.href = window.URL.createObjectURL(blob, {
+        type: exportExtensionMap[extension],
+      });
+      downloadButton.target = '_blank';
+      let now = new Date();
+      let filename = this.exportStrings('downloadFilename')
+        .replace('{year}', now.getFullYear())
+        .replace('{month}', now.toLocaleString('default', { month: 'long' }));
+      downloadButton.download = `${filename}.${extension}`;
+      downloadButton.click();
+    },
+    downloadChannelsCSV(query) {
+      const headers = [
+        this.exportStrings('id'),
+        this.exportStrings('name'),
+        this.exportStrings('description'),
+        this.exportStrings('language'),
+        this.exportStrings('token'),
+        this.exportStrings('size'),
+        this.exportStrings('storage'),
+        this.exportStrings('resources'),
+        this.exportStrings('languages'),
+        this.exportStrings('subtitles'),
+        this.exportStrings('coachContent'),
+        this.exportStrings('assessments'),
+        this.exportStrings('tags'),
+        this.exportStrings('authors'),
+        this.exportStrings('providers'),
+        this.exportStrings('aggregators'),
+        this.exportStrings('licenses'),
+        this.exportStrings('copyrightHolders'),
+      ];
+
+      return this.getChannelListDetails(query).then(channelList => {
+        let csv = Papa.unparse({
+          fields: headers,
+          data: channelList.map(channel => [
+            channel.id,
+            channel.name,
+            channel.description,
+            this.translateLanguage(channel.language),
+            channel.primary_token
+              ? `${channel.primary_token.slice(0, 5)}-${channel.primary_token.slice(5, 10)}`
+              : '',
+            this.$formatNumber(channel.resource_count),
+            this.formatFileSize(channel.resource_size),
+            sortBy(channel.kind_count, 'kind_id')
+              .map(kind =>
+                this.exportStrings('kindCount')
+                  .replace('{kind}', this.translateConstant(kind.kind_id))
+                  .replace('{count}', this.$formatNumber(kind.count))
+              )
+              .join(' • '),
+            channel.languages.join(' • '),
```

Comment:
Hrm, I think maybe just skip the join here, and pass the array to papaparse? In general it should wrap it in escape chars `""` and then comma separate inside it.

Reply:
For i18n, are the commas something we'd want to avoid?

Reply:
Any sort of automated join in this way is something we'd want to avoid, whether comma or bullet. In the CSV case, this is about the best we can do without doing something much more clever. (e.g. using the IntlListFormat, which unfortunately is not widely supported by browsers: https://github.com/wessberg/intl-list-format)

Reply:
Opened https://github.com/learningequality/studio/issues/1884 in the meantime

Reply:
FYI here is one strategy we use in Kolibri: https://github.com/learningequality/kolibri/blob/2596721944a555b7a7b6d2936b05f2e299a163d3/kolibri/plugins/coach/assets/src/views/common/TruncatedItemList.vue

It's also possible (if a little hacky) to do a semantically-correct list like:

```html
<ul>
  <li>item 1</li>
  <li>item 2</li>
  <li>item 3</li>
</ul>
```

```css
ul {
  padding: 0;
  list-style-type: none;
}

li {
  display: inline-block;
}

li+li:before {
  content: "•";
  margin-right: 6px;
}
```

which renders

![image](https://user-images.githubusercontent.com/2367265/82000596-babe5b00-960d-11ea-9ad4-9d9eff023836.png)